### ISSUE TITLE
java.lang.StackOverflowError: SqlInsertBuilder.kt:368

### ISSUE THREAD

**Issue by stefanos-kalantzis** [EMResearch/EvoMaster#240]
(follow up of #239)

Getting a StackOverflowError when running `java -jar evomaster.jar`.

```
19:06:27.559 [main] WARN  o.e.c.r.service.RemoteController - Error message: Failed to execute database command: Failed to execute insertion with index 0 with SQL: INSERT INTO [...] Unique index or primary key violation [...]
* Consumed search budget: 23.503%; covered targets: 393; time per test: 44.8ms (1.3 actions)
Exception in thread "main" java.lang.StackOverflowError                                                                                                                                                                                                                                                                        
        at org.evomaster.core.database.schema.Column.hashCode(Column.kt)                                                                                                                                                                                                                                                       
        at java.base/java.util.HashMap.hash(HashMap.java:339)                                                                                                                                                                                                                                                                  
        at java.base/java.util.HashMap.containsKey(HashMap.java:591)
        at java.base/java.util.HashSet.contains(HashSet.java:204)
        at org.evomaster.core.database.DbActionGeneBuilder.getForeignKey(DbActionGeneBuilder.kt:22)
        at org.evomaster.core.database.DbActionGeneBuilder.buildGene(DbActionGeneBuilder.kt:27)
        at org.evomaster.core.database.DbAction.<init>(DbAction.kt:55)
        at org.evomaster.core.database.DbAction.<init>(DbAction.kt:31)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:360)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
        at org.evomaster.core.database.SqlInsertBuilder.createSqlInsertionAction(SqlInsertBuilder.kt:368)
[...]
        at org.evomaster.core.problem.rest.service.RestSampler.sampleSqlInsertion(RestSampler.kt:44)
        at org.evomaster.core.problem.rest.service.RestStructureMutator.addInitializingActions(RestStructureMutator.kt:68)
        at org.evomaster.core.search.service.mutator.Mutator.mutateAndSave(Mutator.kt:107)
        at org.evomaster.core.search.algorithms.MioAlgorithm.search(MioAlgorithm.kt:51)
        at org.evomaster.core.Main$Companion.run(Main.kt:290)
        at org.evomaster.core.Main$Companion.initAndRun(Main.kt:151)
        at org.evomaster.core.Main$Companion.main(Main.kt:73)
        at org.evomaster.core.Main.main(Main.kt)
```

**Comment by stefanos-kalantzis** [comment]
After looking at the code a bit, I found [this piece](https://github.com/EMResearch/EvoMaster/blob/master/core/src/main/kotlin/org/evomaster/core/database/SqlInsertBuilder.kt#L363):
```
        for (fk in table.foreignKeys) {
            /*
                Assumption: in a valid Schema, this should never end up
                in a infinite loop?
             */
            val pre = createSqlInsertionAction(fk.targetTable, setOf())
            actions.addAll(0, pre)
        }
```

Unfortunately this doesn't work if you're saving a **tree** in an RDBMS, because naturally you must have a self reference:

```
@Entity
public class TreeNode {

    @ManyToOne
    private TreeNode parent;
}
```

**Comment by arcuri82** [comment]
Indeed! self-references is something we should had thought about... 

for now you can try to disable SQL support. the related [options](https://github.com/EMResearch/EvoMaster/blob/master/docs/options.md) should be: `--extractSqlExecutionInfo false --generateSqlDataWithSearch false --heuristicsForSQL false`

**Comment by arcuri82** [comment]
Hi,
 I made a fix in the master branch. if you build `evomaster.jar`, could you check if you still see this error? thx

**Comment by arcuri82** [comment]
this should now be fixed in the new 1.2.0 release. if not, please open a new issue
